src/test/scala/com/cognite/sdk/scala/common/OAuth2SessionTest.scala

```diff
@@ -0,0 +1,122 @@
+package com.cognite.sdk.scala.common
+
+import cats.effect._
+import cats.effect.laws.util.TestContext
+import cats.implicits.catsStdInstancesForList
+import cats.syntax.parallel._
+import com.cognite.sdk.scala.v1.SessionTokenResponse
+import org.scalatest.OptionValues
+import org.scalatest.flatspec.AnyFlatSpec
+import org.scalatest.matchers.should.Matchers
+import sttp.client3._
+import sttp.client3.impl.cats.implicits._
+import sttp.client3.asynchttpclient.cats.AsyncHttpClientCatsBackend
+import sttp.client3.testing.SttpBackendStub
+import sttp.model.{Header, MediaType, Method, StatusCode}
+import sttp.monad.MonadError
+
+import java.util.concurrent.Executors
+import scala.collection.immutable.Seq
+import scala.concurrent.ExecutionContext
+import scala.concurrent.duration._
+
+@SuppressWarnings(Array("org.wartremover.warts.Var"))
+class OAuth2SessionTest extends AnyFlatSpec with Matchers with OptionValues {
+
+  implicit val testContext: TestContext = TestContext()
+  implicit val cs: ContextShift[IO] =
+    IO.contextShift(ExecutionContext.fromExecutor(Executors.newFixedThreadPool(4)))
+  implicit val timer: Timer[IO] = testContext.timer[IO]
+
+  // Override sttpBackend because this doesn't work with the testing backend
+  implicit val sttpBackend: SttpBackend[IO, Any] = AsyncHttpClientCatsBackend[IO]().unsafeRunSync()
+
+  it should "refresh tokens when they expire" in {
+    import sttp.client3.impl.cats.implicits._
+
+    val projectName = "irrelevant"
+    var numTokenRequests = 0
+
+    implicit val mockSttpBackend: SttpBackendStub[IO, Any] =
+      SttpBackendStub(implicitly[MonadError[IO]])
+        .whenRequestMatches { req =>
+          req.method === Method.POST && req.uri.path.endsWith(
+            Seq(projectName, "sessions", "token")
+          ) &&
+          req.body === StringBody(
+            """{"sessionKey":"sessionKey-value"}""",
+            "utf-8",
+            MediaType.ApplicationJson
```

Comment:
Please check the vaultToken as well!

Reply:
sure, done in https://github.com/cognitedata/cognite-sdk-scala/pull/401/commits/b8fbbc2492c7155e30034ca1b8c644640afef1fd ;)